# Prelecture Homework
## 1. 

In [3]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)
df.isna().sum()

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

## 2. Chatbot session started
### 2. i. 

In [10]:
print(df.shape)
df

(391, 11)


row_n        id      name  gender    species birthday personality  \
0        2   admiral   Admiral    male       bird     1-27      cranky   
1        3   agent-s   Agent S  female   squirrel      7-2       peppy   
2        4     agnes     Agnes  female        pig     4-21        uchi   
3        6        al        Al    male    gorilla    10-18        lazy   
4        7   alfonso   Alfonso    male  alligator      6-9        lazy   
..     ...       ...       ...     ...        ...      ...         ...   
386    475    winnie    Winnie  female      horse     1-31       peppy   
387    477  wolfgang  Wolfgang    male       wolf    11-25      cranky   
388    480      yuka      Yuka  female      koala     7-20      snooty   
389    481      zell      Zell    male       deer      6-7        smug   
390    483    zucker    Zucker    male    octopus      3-8        lazy   

                song    phrase            full_id  \
0         Steep Hill   aye aye   villager-admiral   
1            DJ K.K.  sidekick   villager-agent-s   
2         K.K. House   snuffle     villager-agnes   
3         Steep Hill   Ayyeeee        villager-al   
4        Forest Life  it'sa me   villager-alfonso   
..               ...       ...                ...   
386         My Place    hay-OK    villager-winnie   
387        K.K. Song   snarrrl  villager-wolfgang   
388     Soulful K.K.   tsk tsk      villager-yuka   
389         K.K. D&B     pronk      villager-zell   
390  Spring Blossoms     bloop    villager-zucker   

                                                   url  
0    https://villagerdb.com/images/villagers/thumb/...  
1    https://villagerdb.com/images/villagers/thumb/...  
2    https://villagerdb.com/images/villagers/thumb/...  
3    https://villagerdb.com/images/villagers/thumb/...  
4    https://villagerdb.com/images/villagers/thumb/...  
..                                                 ...  
386  https://villagerdb.com/images/villagers/thumb/...  
387  https://villagerdb.com/images/villagers/thumb/...  
388  https://villagerdb.com/images/villagers/thumb/...  
389  https://villagerdb.com/images/villagers/thumb/...  
390  https://villagerdb.com/images/villagers/thumb/...  

[391 rows x 11 columns]

### 2. ii. Broad descriptions of variables, observation, 
 

### Individual rows or data points in a dataset are called observations. Every row in this dataset represents one villager. 
### A dataset's properties and columns are its variables. Every variable in this dataset corresponds to an observational characteristic of the locals. They give an explanation of the measurements used for the observations.
 

# 3.

In [16]:
df.info

<bound method DataFrame.info of      row_n        id      name  gender    species birthday personality  \
0        2   admiral   Admiral    male       bird     1-27      cranky   
1        3   agent-s   Agent S  female   squirrel      7-2       peppy   
2        4     agnes     Agnes  female        pig     4-21        uchi   
3        6        al        Al    male    gorilla    10-18        lazy   
4        7   alfonso   Alfonso    male  alligator      6-9        lazy   
..     ...       ...       ...     ...        ...      ...         ...   
386    475    winnie    Winnie  female      horse     1-31       peppy   
387    477  wolfgang  Wolfgang    male       wolf    11-25      cranky   
388    480      yuka      Yuka  female      koala     7-20      snooty   
389    481      zell      Zell    male       deer      6-7        smug   
390    483    zucker    Zucker    male    octopus      3-8        lazy   

                song    phrase            full_id  \
0         Steep Hill   aye

## 3. 

In [17]:
df.describe()

row_n
count  391.000000
mean   239.902813
std    140.702672
min      2.000000
25%    117.500000
50%    240.000000
75%    363.500000
max    483.000000

In [25]:
df['id'].value_counts()

id
admiral    1
mott       1
paula      1
patty      1
pate       1
          ..
eloise     1
elmer      1
ellie      1
elise      1
zucker     1
Name: count, Length: 390, dtype: int64

In [27]:
df.isna().sum()

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

In [28]:
df.shape

(391, 11)

# 4.

In [31]:
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
df.isna().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [32]:
df.shape

(891, 15)

In [34]:
df.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

#### In df.shape, it indicates exactly the number of row (observation) elements there are, but in df.describe, it tallied the number of valid entries there were. It didn't count the 177 missing entry values under "age," since those entries are nonexistent. As a result, it tallied the existing values (714).     891 - 177 = 714    Furthermore, although df.describe() only shows six columns, df.shape lists fifteen. This discrepancy arises from df.describe()'s limited ability to compute numerical values, which forces it to make computations for the six columns that contain numerical values. 

# 5.

#### An attribute is a property of an object; as such, it can be accessed without parenthesis. Since they simply store static data, they are incapable of carrying out any computational operations. It provides you with already-existing information.
#### A method, however, carries out a task. They accept inputs and are capable of retrieving, manipulating, and doing calculations. Depending on the action, methods convert available information into something you can use.
 

# Post-lecture homework
# 6.

In [35]:
df.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

#### 'Count' displays the number of non-missing values within each column. It shows how many actual data points there are in a given column
#### 'Mean' is the average of the values in a given column. It is calculated by adding all of the values and dividing the total by the count of values for the column displayed under 'count'. 
#### 'std', meaning standard deveation, is a measure of how spread out the values are around the mean. The higher the standard deviation, the more the values are spread out, and the lower it is, the more the values are clustered around the mean.
#### 'min', meaning minimum, is the smallest value in the column. 
#### '25%' This is the 25th percentile of the data, meaning 25% of the values in the column are below this value. It is also known as the lower quartile.
#### '50%', The 50th percentile, also known as the median, is the middle value when the data points are arranged in order. Half of the values are below the median and half are above.
#### '75%', The 75th percentile, meaning 75% of the values in the column are below this value.
#### 'max', meaning maximum is the biggest value in the column.

# 7.

###The number of non-missing values in each column is shown with the command "Count." It displays the number of real data points in a particular column. The average of the values in a particular column is called "mean." To calculate it, sum up all the values, then divide the total by the number of items in the column that is labelled "count." The standard deviation, or "std," represents the degree to which the data deviate from the mean. The values are more dispersed the larger the standard deviation, and more concentrated around the mean the smaller it is. The value in the column that is the least is called "min," or minimum.
"25 percent" This represents 25% of the values in the data, or the 25th percentile. 


In [4]:
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
df.isna().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

#### Before: As shown above, before you drop the missing values, you can see that there are 177 missing values under 'age' and 688 missing values under 'deck'

In [44]:
df.dropna()
df=df.dropna()

In [53]:
df.isna().sum()

survived       0
pclass         0
sex            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
deck           0
embark_town    0
alive          0
alone          0
dtype: int64

In [48]:
print(df.shape)

(182, 15)


#### After: I used df.dropna, because i'm not sure if 'age' and/or 'deck' are important columns for an analysys. When using df.dropna, it dropped all the rows with missing values, thus leaving me with a dataset with 182 rows and 15 columns. I didn't have to drop any potentially important column information that may be useful for analysis, while still having enough rows of data for the information to be useful. 

# 8.

In [59]:
df.columns

Index(['survived', 'pclass', 'sex', 'sibsp', 'parch', 'fare', 'embarked',
       'class', 'who', 'adult_male', 'deck', 'embark_town', 'alive', 'alone'],
      dtype='object')

In [64]:
df.groupby("survived")["fare"].describe()

count       mean        std   min    25%      50%   75%       max
survived                                                                   
0          59.0  64.532131  62.108090  0.00  27.15  50.4958  79.2  263.0000
1         123.0  85.821107  81.843522  8.05  30.50  69.3000  93.5  512.3292

#### Using the.describe() method, it categorized survivors from non-survivors and counted multiple metrics of the fare connected to each of the two options (survivor/non-survivor).

In [66]:
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
df

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

#### There are differences in the values of df.groupby("col1")["col2"] and df.describe().due to df.groupby("col1")["col2"]. description().describe() is unique to the data in those specified columns with respect to another column (grouped by values of specific columns). In contrast, df.describe() is non-specific and applies to the entire df.

In [67]:
df.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

# 8.3

In [1]:
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)

NameError: name 'pd' is not defined

In [2]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)

####  A) 

In [5]:
url = "titanics.csv"
df = pd.read_csv(url)
df.isna().sum()

FileNotFoundError: [Errno 2] No such file or directory: 'titanics.csv'

#### Although it indicated that a URL was missing, I still had to utilize Google to look up a real link to enter as the URL.

In [1]:
dF.Shape

NameError: name 'dF' is not defined

###Chat suggrst that I might've accidentally used dF insread of df, which is correct because I purposely mispelled df into DF


In [2]:
DF.groupby("pclass")["age"].describe()


NameError: name 'DF' is not defined

##the DF is not defined therefore the function cannot run because the code is trying to run my variable DF, but it is not defined. 

Question 8) 3d

In [6]:
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url
df.isna().sum()

SyntaxError: '(' was never closed (4254815826.py, line 2)

###the syntax error tells me that the '(' was never closed, meaning the bracket around (url) needed to be closed but infact wasn't which lead to the code ultimately not being able to run.

In [ ]:
###Question 3e)

In [9]:
import pandas as pd
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)


In [10]:

df.despicable()


AttributeError: 'DataFrame' object has no attribute 'despicable'

###the datafram tells me that "despicable has no attribute which is true because I changed the function from describe to despicable to see if the system would be able to catch that mistake and it did so accuratley while naming the mistake as well"

In [13]:
df.groupby("Sex")["age"].describe()

KeyError: 'Sex'

###the key error in this was naming the variable sex with an uppercase S which isn't defined in the system. Therefore it is a unamed variable.

In [14]:
titanic_df.groupby("sex")[age].describe()

NameError: name 'titanic_df' is not defined

###titanic_df isn't defined therefore chatgpt noticed the problem right away and show the reader that the variable titanic-df is not defined. 

###Here's a summary of everything we discussed tonight:

Titanic Dataset:

You loaded the Titanic dataset from a URL using pd.read_csv().
We explored missing data in the dataset using df.isna().sum(), which revealed that columns like age and deck have missing values.
Descriptive Statistics:

You used df.describe() to generate summary statistics (count, mean, standard deviation, min, 25%, 50%, 75%, and max) for each column.
We discussed how these statistics provide insights into the distribution and spread of numeric data in the DataFrame.
Difference Between df.shape and df.describe():

df.shape gives the total number of rows and columns in the dataset, whereas df.describe() gives statistical summaries, which only count non-null values. Missing values will cause the count in df.describe() to be less than the row count in df.shape.
Difference Between Attributes and Methods:

Attributes (like df.shape) do not require parentheses and refer to characteristics of the DataFrame.
Methods (like df.describe()) require parentheses and perform actions on the DataFrame, such as calculating statistics.
Use Case for df.dropna() vs del df['col']:

df.dropna() is useful when you want to remove rows or columns that contain missing values, while keeping the overall structure of the dataset intact.
del df['col'] is preferred when you want to permanently remove an entire column from the dataset, even if there are no missing values.
Combination of del df['col'] and df.dropna():

We discussed how applying del df['col'] before df.dropna() might be important to remove irrelevant columns before cleaning the dataset. This helps avoid inadvertently dropping too many rows if the dropped column had excessive missing data.
df.groupby("col1")["col2"].describe():

We explored the groupby method in Python, which allows you to group the data by one column (col1) and calculate summary statistics for another column (col2). We used df.groupby("pclass")["age"].describe() to see age statistics for each passenger class in the Titanic dataset.
NameError: Name 'DF' is Not Defined:

We discussed how Python is case-sensitive and how attempting to use DF instead of df (or vice versa) leads to a NameError because Python cannot find a variable called DF if it's actually defined as df.
In summary, we covered important concepts related to data cleaning, handling missing values, exploring group-wise statistics, and the importance of consistent variable naming in Python.









